### DISCUSSION TITLE
GPT recursion

### DISCUSSION THREAD

**Discussion by mpoon** [mpoon/gpt-repository-loader#18] [Announcements]
I was getting tired of copy/pasting reams of code into GPT-4 to give it context before I asked it to help me, so I started this small tool. In a nutshell, `gpt-repository-loader` will spit out file paths and file contents in a prompt-friendly format. You can also use `.gptignore` to ignore files/folders that are irrelevant to your prompt.

`gpt-repository-loader` as-is works pretty well in helping me achieve better responses. Eventually, I thought it would be cute to load itself into GPT-4 and have GPT-4 improve it. I was honestly surprised by [PR#17](https://github.com/mpoon/gpt-repository-loader/pull/17/). GPT-4 was able to write a valid an example repo and an expected output and throw in a small curveball by adjusting `.gptignore`. I did tell GPT the output file format in two places: 1.) in the preamble when I prompted it to make a PR for [issue #16](https://github.com/mpoon/gpt-repository-loader/issues/16) and 2.) as a string in `gpt_repository_loader.py`, both of which are indirect ways to infer how to build a functional test. However, I don't think I explained to GPT in English anywhere on how `.gptignore` works at all!

I wonder how far GPT-4 can take this repo. Here is the process I'm following for developing:

1. Open an issue describing the improvement to make
2. Construct a prompt - start with using `gpt_repository_loader.py` on this repo to generate the repository context, then append the text of the opened issue after the `--END--` line.
3. Try not to edit any code GPT-4 generates. If there is something wrong, continue to prompt GPT to fix whatever it is.
4. Create a feature branch on the issue and create a pull request based on GPT's response.
5. Have a maintainer review, approve, and merge.

I am going to try to automate the steps above as much as possible. Really curious how tight the feedback loop will eventually get before something breaks!

**Comment by zackees** [comment]
I used my toolset to package up your codebase.

It now installs a command into the system path..

```
pip install gptrepo
gptrepo <path>
```

Also added automated testing and linting. It builds and runs on Mac/Windows/Linux.

Check it out:
https://github.com/zackees/gptrepo

**Reply by cachho** [reply]
can we merge this or some? You're some behind, some ahead. 

But I'm also having the problem `.venv/src/gpt-repository-loader does not appear to be a Python project: no pyproject.toml or setup.py`

**Comment by cur8tor** [comment]
what is the exact prompt you are typing into chat gpt after generating the output.txt file? Isn't the contents of the output far to long for chat gpt to take as an input? I am having this issue.

**Reply by zackees** [reply]
Only extremely small repo's work right now. You'll have to wait until gpt4-32k comes which allows 52 pages of text.

**Reply by truimagz** [reply]
GPT starts to hallucinate very quickly about having file contents you never gave it

**Reply by x0us** [reply]
> Only extremely small repo's work right now. You'll have to wait until gpt4-32k comes which allows 52 pages of text.

Hi, Guys, any updates for the 32k token launch date?

**Reply by notamicrodose** [reply]
> > Only extremely small repo's work right now. You'll have to wait until gpt4-32k comes which allows 52 pages of text.
> 
> Hi, Guys, any updates for the 32k token launch date?

I read that it should be out in a couple weeks but i dont think that it's confirmed

**Reply by svallory** [reply]
It's out already but you have to join the waitlist and wait until they let you in
